File: problem-types/lamps-play.py
import json

def entry_form(data, kwargs):
	w, h = data['width'], data['height']
	switches = data['switches']
	yield '<table class="grid">'
	for y in range(h):
		yield '<tr>'
		for x in range(w):
			yield f'<td id="lamp_{x}_{y}" class="lamp lamp_on">'
	yield '</table>'
	yield '<div class="lamp_bar">'
	for k, switch in enumerate(switches):
		yield '<div>'
		yield '<table class="grid">'
		assert len(switch) == h
		for y, row in enumerate(switch):
			assert len(row) == w
			yield '<tr>'
			for x, val in enumerate(row):
				classes = 'lamp'
				if val:
					classes += ' lamp_on'
				yield f'<td id="lamp_{x}_{y}" class="{classes}">'
		yield '</table>'
		yield f'<button type="button" class="lamp_switch" data-id="{k}">{k + 1}</button>'
		yield '</div>'
	yield '</div>'
	yield f"""
		<script type="text/ecmascript">
			"use strict"
			const w = {w}
			const h = {h}
			const switches = {json.dumps(switches)}
		</script>
		"""
	yield f'<form method="post" id="problem_form" class="button_bar">'
	yield '<input id="i_answer" type="hidden" name="answer" >'

def validate(data, answer):
	answer = json.loads(answer)
	return answer in data['correct']

File: problem-types/lamps-play.js
"use strict"

var buttons = document.querySelectorAll('.lamp_switch');

for (let button of buttons){
	button.onclick = function(){
		if (button.classList.contains('active'))
			button.classList.remove('active');
		else
			button.classList.add('active');
	}
}
document.addEventListener("DOMContentLoaded", () => {
	let answer = Array(switches.length).fill(0)

	function toggle_switch(id) {
		answer[id] ^= 1
		let sw = switches[id]
		for (let y = 0; y < h; y++)
			for (let x = 0; x < w; x++) {
				if (sw[y][x])
					document.getElementById(`lamp_${x}_${y}`).classList.toggle('lamp_on')
			}
		i_answer.value = JSON.stringify(answer)
	}

	for (let sw of document.getElementsByClassName("lamp_switch")) {
		let id = sw.dataset.id
		sw.addEventListener("click", (e) => toggle_switch(id))
	}
})
